source/Mlos.Python/mlos/Optimizers/RegressionModels/SklearnRandomForestRegressionModel.py

```diff
@@ -0,0 +1,229 @@
+        is set to "mse", then `std[i] ~= std(y | X[i])`.
+    """
+    # This derives std(y | x) as described in 4.3.2 of arXiv:1211.0906
+    std = np.zeros(len(X))
+
+    for tree in trees:
+        var_tree = tree.tree_.impurity[tree.apply(X)]
+
+        # This rounding off is done in accordance with the
+        # adjustment done in section 4.3.3
+        # of http://arxiv.org/pdf/1211.0906v2.pdf to account
+        # for cases such as leaves with 1 sample in which there
+        # is zero variance.
+        var_tree[var_tree < min_variance] = min_variance
+        mean_tree = tree.predict(X)
+        std += var_tree + mean_tree ** 2
+
+    std /= len(trees)
+    std -= predictions ** 2.0
+    std[std < 0.0] = 0.0
+    std = std ** 0.5
+    return std
+
+class SklearnRandomForestRegressionModel(RegressionModel):
+    """Thin wrapper for the scikit-learn RandomForestRegressor.
+
+    This model natively support multi-output regression.
+    """
+
+    _PREDICTOR_OUTPUT_COLUMNS = [
+        Prediction.LegalColumnNames.IS_VALID_INPUT,
+        Prediction.LegalColumnNames.PREDICTED_VALUE,
+        Prediction.LegalColumnNames.PREDICTED_VALUE_VARIANCE,
+        Prediction.LegalColumnNames.PREDICTED_VALUE_DEGREES_OF_FREEDOM
+    ]
+
+    def __init__(
+            self,
+            model_config: Point,
+            input_space: Hypergrid,
+            output_space: Hypergrid,
+            logger: logging.Logger = None
+    ):
+        if logger is None:
+            logger = create_logger("SklearnRandomForestRegressionModel")
+        self.logger = logger
+        assert model_config in SklearnRandomForestRegressionModelConfig.CONFIG_SPACE
+
+        if model_config.oob_score:
+            model_config.bootstrap = True
+
+        if len(output_space.dimensions) > 1:
+            ValueError("Multi-output regression is not yet supported. ")
+
+        RegressionModel.__init__(
+            self,
+            model_type=type(self),
+            model_config=model_config,
+            input_space=input_space,
+            output_space=output_space
+        )
+
+        self.model_config = model_config
+
+        # one hot encode categorical input dimensions
+        self.one_hot_encoder_adapter = CategoricalToOneHotEncodedHypergridAdapter(
+            adaptee=self.input_space,
+            merge_all_categorical_dimensions=False)
+
+        self.input_dimension_names = [dimension.name for dimension in self.input_space.dimensions]
+        self._projected_input_dimension_names = [dimension.name for dimension in self.one_hot_encoder_adapter.dimensions]
+        self.continuous_dimension_names = [dimension.name for dimension in self.one_hot_encoder_adapter.target.dimensions
+                                           if isinstance(dimension, ContinuousDimension)]
+        self.output_dimension_names = [dimension.name for dimension in self.output_space.dimensions]
+
+        self.random_forest_regressor_ = None
+
+        self.fit_X_ = None

```

Comment:
Unused?

Reply:
Actually a couple of these look like that?  Or maybe they're elsewhere in the hierarchy and I'm just misremembering.